port/unix/omrsl.c

```diff
@@ -438,3 +438,97 @@ omrsl_startup(struct OMRPortLibrary *portLibrary)
 {
 	return 0;
 }
+
+/**
+ * Native Libraries
+ *
+ * This function is called go get all the shared libraries loaded by a process.
+ *
+ * @param[in] portLibrary Pointer to the OMR port library.
+ * @param[in] callback Function to be called for each library found.
+ * @param[in] userData User-defined data passed to the callback.
+ *
+ * @return 0 if successful, or the first non-zero return value from the callback.
+ */
+uintptr_t
+omrsl_get_libraries(struct OMRPortLibrary *portLibrary, OMRLibraryInfoCallback callback, void *userData)
+{
+#if defined(LINUX)
+	#include <limits.h>
+	/* Length of buffer. */
+	#define READ_CHUNK_SIZE (PATH_MAX + 100)
+	char buffer[READ_CHUNK_SIZE + 1];
+	char carryOver[READ_CHUNK_SIZE + 1];
+	intptr_t bytesRead = 0;
+	uintptr_t result = 0;
+	void *addrLow = NULL;
+	void *addrHigh = NULL;
+	unsigned long offset = 0;
+	int devMajor = 0;
+	int devMinor = 0;
+	unsigned long inode = 0;
+	char permissions[5] = {0};
+	char path[READ_CHUNK_SIZE];
+	int count = 0;
+	char combinedBuffer[2 * READ_CHUNK_SIZE + 1];
+	int lastCharIsNewline = 0;
+	char *line = NULL;
+	carryOver[0] = '\0';
+	intptr_t fd = portLibrary->file_open(portLibrary, "/proc/self/maps", EsOpenRead, 0);
+	if (fd < 0) {
+		int portableError = portLibrary->error_last_error_number(portLibrary);
+		Trc_PRT_find_executable_name_failedOpeningProcFS(portableError);
+		portLibrary->error_set_last_error_with_message(portLibrary, portableError, "Failed to open /proc fs");
+		return -1;
+	}
+	do {
+		bytesRead = portLibrary->file_read(portLibrary, fd, buffer, sizeof(buffer) - 1);
+		if (bytesRead <= 0) {
+			break;
+		}
+		buffer[bytesRead] = '\0';
+		lastCharIsNewline = '\n' == buffer[bytesRead - 1];
+		combinedBuffer[0] = '\0';
+		portLibrary->str_printf(portLibrary, combinedBuffer, sizeof(combinedBuffer), "%s%s", carryOver, buffer);
+		line = strtok(combinedBuffer, "\n");
+		while (NULL != line) {
+			char *nextLine = strtok(NULL, "\n");
+			if ((NULL == nextLine) && (0 == lastCharIsNewline)) {
+				int readChars = portLibrary->str_printf(portLibrary, carryOver, sizeof(carryOver), "%s", line);
+				if(readChars >= sizeof(carryOver)) {
+					int portableError = portLibrary->error_last_error_number(portLibrary);
+					portLibrary->error_set_last_error_with_message(portLibrary, portableError, "String truncation occurred");
+				}
+				break;
+			} else {
+				carryOver[0] = '\0';
+			}
+			addrLow = NULL;
+			addrHigh = NULL;
+			offset = 0;
+			devMajor = 0;
+			devMinor = 0;
+			inode = 0;
+			memset(permissions, 0, sizeof(permissions));
+			memset(path, 0, sizeof(path));
+			count = sscanf(
+					line,
+					"%p-%p %4s %lx %x:%x %lu %s",
+					&addrLow, &addrHigh, permissions,
+					&offset, &devMajor, &devMinor, &inode, path);
+			if ((8 == count)  && ('/' == path[0])) {
+				result = callback(path, addrLow, addrHigh, userData);
+				if (0 != result) {
+					break;
+				}
+			}
+			line = nextLine;
+		}
+	} while (bytesRead < 0);
```

Comment:
`bytesRead` cannot be negative here (see lines 486-488); please replace the "do...while" loop with a `for (;;)` loop.